source/game/system/KPadController.cc

```diff
@@ -0,0 +1,229 @@
+    buffer += directionCount;
+
+    m_buttonsStreams[2]->buffer = buffer;
+    m_buttonsStreams[2]->size = trickCount * SEQUENCE_SIZE;
+}
+
+void KPadGhostController::calcInner() {
+    if (!m_ghostBuffer || !m_acceptingInputs) {
+        return;
+    }
+
+    m_raceInputState.buttons = m_buttonsStreams[0]->readFrame() & 0xFF;
+    u16 sticks = m_buttonsStreams[1]->readFrame();
+    m_raceInputState.stickXRaw = sticks >> 4 & 0xF;
+    m_raceInputState.stickYRaw = sticks & 0xF;
+    m_raceInputState.stick =
+            EGG::Vector2f((static_cast<f32>(m_raceInputState.stickXRaw) - 7.0f) / 7.0f,
+                    (static_cast<f32>(m_raceInputState.stickYRaw) - 7.0f) / 7.0f);
+    m_raceInputState.trickRaw = m_buttonsStreams[2]->readFrame() >> 0x4;
+    m_raceInputState.trick = m_raceInputState.trickRaw;
+
+    K_LOG("Buttons: %X\nSticks: %X\nTrick: %X", m_raceInputState.buttons, sticks,
+            m_raceInputState.trickRaw);
+}
+
+void KPadGhostController::setAcceptingInputs(bool set) {
+    m_acceptingInputs = set;
+}
+
+RaceInputState::RaceInputState() {
+    reset();
+}
+
+void RaceInputState::reset() {
+    buttons = 0;
+    buttonsRaw = 0;
+    stick = EGG::Vector2f::zero;
+    stickXRaw = 7;
+    stickYRaw = 7;
+    trick = 0;
+    trickRaw = 0;
+}
+
+KPadGhostButtonsStream::KPadGhostButtonsStream()
+    : buffer(nullptr), sequenceCount(-1), size(0x276C), state(2) {}
+
+KPadGhostButtonsStream::~KPadGhostButtonsStream() = default;
+
+bool KPadGhostButtonsStream::readIsNewSequence(const u16 *sequence) const {
+    return readSequenceFrames >= (*sequence >> 0x8);
+}
+
+u16 KPadGhostButtonsStream::readVal(const u16 *sequence) const {
+    return *sequence;
+}
+
+u16 KPadGhostButtonsStream::readFrame() {
+    constexpr u32 SEQUENCE_SIZE = 0x2;
+
+    if (state != 1) {
+        return 0;
+    }
+
+    bool bVar1 = false;
+    const u16 *sequence = nullptr;
+
+    if (sequenceCount >= 0) {
+        sequence = buffer + sequenceCount;
+
+        if (readIsNewSequence(sequence)) {
+            bVar1 = true;
+            if (size <= ++sequenceCount * SEQUENCE_SIZE) {
+                state = 2;
+            }
+        }
+    } else {
+        sequenceCount = 0;
+        bVar1 = true;
+    }
+
+    if (state != 1) {
+        return 0;
+    }
+
+    sequence = buffer + sequenceCount;
+
+    if (bVar1) {
+        readSequenceFrames = 0;
+    }
+
+    ++readSequenceFrames;
+
+    return readVal(sequence);
+}
+
+KPadGhostFaceButtonsStream::KPadGhostFaceButtonsStream() : KPadGhostButtonsStream() {}
+
+KPadGhostFaceButtonsStream::~KPadGhostFaceButtonsStream() = default;
+
+KPadGhostDirectionButtonsStream::KPadGhostDirectionButtonsStream() : KPadGhostButtonsStream() {}
+
+KPadGhostDirectionButtonsStream::~KPadGhostDirectionButtonsStream() = default;
+
+KPadGhostTrickButtonsStream::KPadGhostTrickButtonsStream() : KPadGhostButtonsStream() {}
+
+KPadGhostTrickButtonsStream::~KPadGhostTrickButtonsStream() = default;
```

Comment:
Since `KPadGhostButtonsStream` has a default constructor, can these constructors be `= default`?

Reply:
Yup!